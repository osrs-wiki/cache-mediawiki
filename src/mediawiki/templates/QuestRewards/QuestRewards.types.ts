// This file is intentionally empty - no props needed for direct template functions
