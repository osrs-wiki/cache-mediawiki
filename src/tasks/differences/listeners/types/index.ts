import { areaListener } from "./areas";
import { itemListener } from "./items";
import { musicListener } from "./music";
import { npcListener } from "./npcs";
import { objectListener } from "./objects";
import { CacheChangeListener } from "../listeners.types";

export const cacheListeners: CacheChangeListener[] = [
  areaListener,
  itemListener,
  musicListener,
  npcListener,
  objectListener,
];
