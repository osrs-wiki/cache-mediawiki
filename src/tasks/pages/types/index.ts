export * from "./area";
export * from "./item";
export * from "./npc";
export * from "./quest";
export * from "./scenery";
