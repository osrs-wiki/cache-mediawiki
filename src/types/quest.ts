export type QuestDifficulty =
  | "Novice"
  | "Intermediate"
  | "Experienced"
  | "Master"
  | "Grandmaster"
  | "Special";

export type QuestLength =
  | "Very Short"
  | "Short"
  | "Medium"
  | "Long"
  | "Very Long";

export type QuestType = "Quests" | "Miniquests";

export interface Quest {
  id: number;
  sortname: string;
  displayname: string;
  inPrerelease: boolean;
  type: QuestType;
  members: boolean;
  difficulty: QuestDifficulty;
  length: QuestLength;
  location: string;
  releasedate: string;
  series: string;
  seriesno: number;
  seriesnoOverride: number;
  startcoord: string;
  startnpc: number;
  startloc: string;
  mapelement: string;
  questpoints: number;
  unstartedstate: number;
  endstate: number;
  version: number;
  parentQuest: number;
  hasSubquests: boolean;
  requirementStats: string;
  recommendedStats: string;
  requirementQuests: string;
  requirementQuestpoints: number;
  requirementCombat: number;
  recommendedCombat: number;
  requirementCheckSkillsOnStart: boolean;
  requirementsBoostable: boolean;
  speedrun: boolean;
  totalXpAwarded: number;
  prerequisiteDirect: string;
  prerequisiteIndirect: string;
  crCanRecommend: boolean;
  crExperienceProfile: string;
  crRecommendationReason: string;
  crRecommendationReasonIsPrimary: boolean;
  crPointsSkill: number;
  crPointsTransport: number;
  crPointsEquipment: number;
  crPointsArea: number;
  crPointsXpType: string;
  crStarter: boolean;
  fswWorldFirstId: number;
}
