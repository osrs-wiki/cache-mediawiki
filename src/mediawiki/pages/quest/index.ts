import {
  MediaWikiBreak,
  MediaWikiBuilder,
  MediaWikiHeader,
  MediaWikiTemplate,
  MediaWikiText,
} from "@osrs-wiki/mediawiki-builder";

import InfoboxQuestTemplate from "../../templates/InfoboxQuest";
import QuestDetails from "../../templates/QuestDetails";
import QuestRewards from "../../templates/QuestRewards";

import { Quest } from "@/types/quest";

export const questPageBuilder = (quest: Quest): MediaWikiBuilder => {
  const infoboxQuest = InfoboxQuestTemplate(quest);
  const questDetails = QuestDetails(quest);
  const questRewards = QuestRewards(quest);

  const builder = new MediaWikiBuilder();

  builder.addContents([infoboxQuest.build(), new MediaWikiBreak()]);

  // Add quest introduction
  if (quest.displayname) {
    builder.addContents([
      new MediaWikiText(quest.displayname, { bold: true }),
      new MediaWikiText(
        ` is a ${quest.difficulty?.toLowerCase() || ""} quest${
          quest.location ? ` in ${quest.location}` : ""
        }.`
      ),
      new MediaWikiBreak(),
      new MediaWikiBreak(),
    ]);
  }

  // Add quest details
  builder.addContents([
    questDetails,
    new MediaWikiBreak(),
    new MediaWikiBreak(),
  ]);

  // Add walkthrough section
  builder.addContents([
    new MediaWikiHeader("Walkthrough", 2),
    new MediaWikiBreak(),
    new MediaWikiBreak(),
  ]);

  // Add rewards section
  builder.addContents([
    new MediaWikiHeader("Rewards", 2),
    questRewards,
    new MediaWikiBreak(),
    new MediaWikiBreak(),
  ]);

  // Add transcript section
  builder.addContents([
    new MediaWikiHeader("Transcript", 2),
    new MediaWikiTemplate("Hastranscript"),
    new MediaWikiBreak(),
  ]);

  // Add quest series navbox (if applicable)
  if (quest.series && quest.displayname) {
    builder.addContents([
      new MediaWikiBreak(),
      new MediaWikiTemplate(quest.displayname),
    ]);
  }

  return builder;
};
