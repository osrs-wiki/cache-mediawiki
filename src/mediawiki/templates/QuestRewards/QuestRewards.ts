import { MediaWikiTemplate } from "@osrs-wiki/mediawiki-builder";

import { Quest } from "@/types/quest";

const QuestRewards = (quest: Quest): MediaWikiTemplate => {
  const params: Record<string, string | number | boolean | undefined> = {
    name: quest.displayname || undefined,
    image: quest.displayname
      ? `[[File:${quest.displayname} rewards scroll.png]]`
      : undefined,
    qp: quest.questpoints > 0 ? quest.questpoints : undefined,
    rewards:
      quest.totalXpAwarded > 0
        ? `Total XP awarded: ${quest.totalXpAwarded.toLocaleString()}`
        : undefined,
  };

  return new MediaWikiTemplate("Quest rewards", params);
};

export default QuestRewards;
