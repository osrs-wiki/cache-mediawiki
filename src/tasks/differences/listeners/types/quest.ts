import { writeQuestPage, questRowToQuest } from "../../../pages/types";
import { CacheChangeListener } from "../listeners.types";

import {
  IndexType,
  ConfigType,
  DBRow,
  DBRowID,
  DBTableID,
  Reader,
} from "@/utils/cache2";

export const questListener: CacheChangeListener = {
  index: IndexType.Configs,
  archive: ConfigType.DbRow,
  handler: async ({ newFile }) => {
    if (!newFile) return;

    const newEntry = DBRow.decode(
      new Reader(newFile.file.data, {
        era: "osrs",
        indexRevision: newFile.index.revision,
      }),
      <DBRowID>newFile.file.id
    );

    // Check if this is a quest from table 0
    if (newEntry && newEntry.table === (0 as DBTableID)) {
      try {
        const quest = questRowToQuest(newEntry);
        if (quest) {
          writeQuestPage(quest);
        }
      } catch (error) {
        console.error(`Error processing quest ${newFile.file.id}:`, error);
      }
    }
  },
};
