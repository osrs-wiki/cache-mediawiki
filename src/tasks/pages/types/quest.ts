import { writePageToFile } from "../pages.utils";

import { questPageBuilder } from "@/mediawiki/pages/quest";
import { Quest, QuestDifficulty, QuestLength, QuestType } from "@/types/quest";
import { CacheProvider, DBTable, DBTableID } from "@/utils/cache2";

export const writeQuestPageFromCache = async (
  cache: Promise<CacheProvider>,
  id: number
) => {
  try {
    // Load quest data from DBTable 0
    const questRows = await DBTable.loadRows(cache, 0 as DBTableID);
    const questRow = questRows.find((row) => row && row.values[0][0] === id);

    if (questRow) {
      const quest = questRowToQuest(questRow);
      if (quest) {
        writeQuestPage(quest);
      }
    }
  } catch (e) {
    console.error(`Error generating page for quest ${id}: `, e);
  }
};

export const writeQuestPage = async (quest: Quest) => {
  const builder = questPageBuilder(quest);
  writePageToFile(builder, "quest", quest.displayname, quest.id.toString());
};

/**
 * Converts a DBRow from the quest table to a Quest object
 */
export function questRowToQuest(row: {
  values: (string | number | bigint | undefined)[][];
}): Quest | null {
  try {
    const values = row.values;

    // Extract values from the row
    const getValue = (index: number) => values[index] && values[index][0];
    const getBooleanValue = (index: number) => Boolean(getValue(index));

    return {
      id: getValue(0) as number,
      sortname: (getValue(1) as string) || "",
      displayname: (getValue(2) as string) || "",
      inPrerelease: getBooleanValue(3),
      type:
        getValue(4) === 0
          ? ("Quests" as QuestType)
          : ("Miniquests" as QuestType),
      members: getBooleanValue(5),
      difficulty: (getValue(6) as QuestDifficulty) || "Novice",
      length: (getValue(7) as QuestLength) || "Short",
      location: (getValue(8) as string) || "",
      releasedate: (getValue(9) as string) || "",
      series: (getValue(10) as string) || "",
      seriesno: (getValue(11) as number) || 0,
      seriesnoOverride: (getValue(12) as number) || 0,
      startcoord: (getValue(13) as string) || "",
      startnpc: (getValue(14) as number) || 0,
      startloc: (getValue(15) as string) || "",
      mapelement: (getValue(16) as string) || "",
      questpoints: (getValue(17) as number) || 0,
      unstartedstate: (getValue(18) as number) || 0,
      endstate: (getValue(19) as number) || 0,
      version: (getValue(20) as number) || 0,
      parentQuest: (getValue(21) as number) || 0,
      hasSubquests: getBooleanValue(22),
      requirementStats: (getValue(23) as string) || "",
      recommendedStats: (getValue(24) as string) || "",
      requirementQuests: (getValue(25) as string) || "",
      requirementQuestpoints: (getValue(26) as number) || 0,
      requirementCombat: (getValue(27) as number) || 0,
      recommendedCombat: (getValue(28) as number) || 0,
      requirementCheckSkillsOnStart: getBooleanValue(29),
      requirementsBoostable: getBooleanValue(30),
      speedrun: getBooleanValue(31),
      totalXpAwarded: (getValue(32) as number) || 0,
      prerequisiteDirect: (getValue(33) as string) || "",
      prerequisiteIndirect: (getValue(34) as string) || "",
      crCanRecommend: getBooleanValue(35),
      crExperienceProfile: (getValue(36) as string) || "",
      crRecommendationReason: (getValue(37) as string) || "",
      crRecommendationReasonIsPrimary: getBooleanValue(38),
      crPointsSkill: (getValue(39) as number) || 0,
      crPointsTransport: (getValue(40) as number) || 0,
      crPointsEquipment: (getValue(41) as number) || 0,
      crPointsArea: (getValue(42) as number) || 0,
      crPointsXpType: (getValue(43) as string) || "",
      crStarter: getBooleanValue(44),
      fswWorldFirstId: (getValue(45) as number) || 0,
    };
  } catch (error) {
    console.error("Error converting quest row to Quest object:", error);
    return null;
  }
}
