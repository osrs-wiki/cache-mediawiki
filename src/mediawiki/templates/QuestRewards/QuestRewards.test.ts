import QuestRewards from "./QuestRewards";

import { Quest, QuestDifficulty, QuestLength, QuestType } from "@/types/quest";
import { NPC } from "@/utils/cache2";

describe("QuestRewards", () => {
  const mockNPC = { id: 123, name: "Test NPC" } as NPC;

  const baseQuest: Quest = {
    id: 170,
    sortname: "The Curse of Arrav",
    displayname: "The Curse of Arrav",
    inPrerelease: false,
    type: "Quests" as QuestType,
    members: true,
    difficulty: "Master" as QuestDifficulty,
    length: "Medium" as QuestLength,
    location: "Uzer Oasis",
    releasedate: "2024-11-06",
    series: "Mahjarrat",
    seriesno: 10,
    seriesnoOverride: 0,
    startcoord: "3505,3038",
    startnpc: mockNPC,
    startloc: "Uzer Oasis",
    mapelement: "",
    questpoints: 2,
    unstartedstate: 0,
    endstate: 1,
    version: 1,
    parentQuest: 0,
    hasSubquests: false,
    requirementStats: "",
    recommendedStats: "",
    requirementQuests: "",
    requirementQuestpoints: 0,
    requirementCombat: 0,
    recommendedCombat: 85,
    requirementCheckSkillsOnStart: false,
    requirementsBoostable: false,
    speedrun: false,
    totalXpAwarded: 120000,
    prerequisiteDirect: "",
    prerequisiteIndirect: "",
    crCanRecommend: false,
    crExperienceProfile: "",
    crRecommendationReason: "",
    crRecommendationReasonIsPrimary: false,
    crPointsSkill: 0,
    crPointsTransport: 0,
    crPointsEquipment: 0,
    crPointsArea: 0,
    crPointsXpType: "",
    crStarter: false,
    fswWorldFirstId: 0,
  };

  it("should render quest rewards", () => {
    const rewards = QuestRewards(baseQuest);
    expect(rewards).toMatchSnapshot();
  });

  it("should render quest rewards without XP", () => {
    const questWithoutXp = { ...baseQuest, totalXpAwarded: 0 };
    const rewards = QuestRewards(questWithoutXp);
    expect(rewards).toMatchSnapshot();
  });

  it("should render miniquest rewards", () => {
    const miniquest = {
      ...baseQuest,
      questpoints: 0,
      type: "Miniquests" as QuestType,
    };
    const rewards = QuestRewards(miniquest);
    expect(rewards).toMatchSnapshot();
  });
});
