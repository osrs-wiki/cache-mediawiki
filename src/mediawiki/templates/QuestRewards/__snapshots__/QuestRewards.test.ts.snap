// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestRewards should render miniquest rewards 1`] = `
MediaWikiTemplate {
  "children": undefined,
  "name": "Quest rewards",
  "options": Object {
    "image": "[[File:The Curse of Arrav rewards scroll.png]]",
    "name": "The Curse of Arrav",
    "rewards": "Total XP awarded: 120,000",
  },
  "params": Array [],
}
`;

exports[`QuestRewards should render quest rewards 1`] = `
MediaWikiTemplate {
  "children": undefined,
  "name": "Quest rewards",
  "options": Object {
    "image": "[[File:The Curse of Arrav rewards scroll.png]]",
    "name": "The Curse of Arrav",
    "qp": 2,
    "rewards": "Total XP awarded: 120,000",
  },
  "params": Array [],
}
`;

exports[`QuestRewards should render quest rewards without XP 1`] = `
MediaWikiTemplate {
  "children": undefined,
  "name": "Quest rewards",
  "options": Object {
    "image": "[[File:The Curse of Arrav rewards scroll.png]]",
    "name": "The Curse of Arrav",
    "qp": 2,
  },
  "params": Array [],
}
`;
